neps/nep-0541.md

```diff
@@ -0,0 +1,196 @@
+    /// It increments for a combination of `signer_id` and `public_key`
+    pub nonce: Nonce,
+    /// Receiver account for this transaction
+    pub receiver_id: AccountId,
+    /// The hash of the block in the blockchain on top of which the given transaction is valid
+    pub block_hash: CryptoHash,
+    /// A list of actions to be applied
+    pub actions: Vec<Action>,
+    /// Optional priority fee (unit is 10^12 yotcoNEAR)
+    pub priority_fee: u64
+}
+```
+
+Here, `priority_fee * 10^12` is the amount of yotcoNEAR attached. We do not use `u128` here because there is no need to have that level of granularity and also save 8 bytes per transaction.
+
+We define the priority of a transaction to be `priority_fee / attached_gas`, where attached gas is the sum of all action costs and attached gas to function call actions.
+Receipts that spawn from a transaction have the same priority as the transaction itself.
+
+Regardless of priority fee, a transaction still needs to pay the base cost, which is determined by `minimum_gas_price`. The current dynamic gas price mechanism will be removed since it does not make sense for the congestion of one shard to affect the gas price of transaction in a completely different shard, even if there is no congestion there.
+
+When a chunk is produced, the transactions should be sorted by priority in descending order. Otherwise the chunk is invalid.
+However, when the priority of two transactions are the same, no specific order is enforced by the protocol.
+
+When a transaction is processed and converted to a receipt, X percent (X is TBD) of the priority fee is burnt and the rest is rewarded to the chunk producer.
+During receipt processing, receipts with highest priorites (from incoming and delayed receipts combined) will get processed first, but up to a fixed amount of gas (the threshold is TBD). The rest of the gas limit is devoted to processing all receipts, regardless of priority, in a FIFO order. This is to ensure that the system does not deadlock in light of [NEP-539](https://github.com/near/NEPs/pull/539).
+
+## Reference Implementation
+
+We discuss the implementation of two important pieces of the proposal, receipt processing and validator reward, in this section.
+
+### Receipt Processing
+
+A max heap is created in the state for storing priorities of delayed receipts. More specifically, we add a trie column `RECEIPTS_PRIORITY_QUEUE: u8 = 13`. Each value in this heap is of the form
+```rust
+struct ReceiptPriority {
+    /// index of the corresponding receipt in the delayed receipts queue. Use u32 here to save some space in the state
+    receipt_index: u32
+    /// priority of the receipt
+    priority: u64
+}
+```
+Since each value is quite small and we don't expect the delayed receipt queue to be extremely long (even if there are 100k delayed receipts, it is only 1.2MB in size), and it only needs to be read and write once per chunk, this heap can be stored as a single value in the state.
```

Comment:
In principle I agree. But one question for stateless validation, do these trie changes end up in the state witness?
Because 1.2MB every chunk does not sound completely trivial.

Reply:
yeah that is a very good point. On the other hand, it does feel to me quite wasteful to store many such small values in the state, which would also result in many state reads.

Reply:
@jakmeier I thought about it a bit more and came up with a different idea. Instead of trying to maintain a heap within the trie, we can use the priority of receipts as trie keys under a new trie column. Since the key is a u64, it means there are at most 16 trie nodes for each value. It makes proof quite easy and works nicely with stateless validation. In addition, it is quite difficult to attack because one needs to attach priority fees to get a new value created, so in practice it would most likely not be a concern. After we switch to stateless validation, this will all be memory accesses so it's definitely not going to be a problem.

Reply:
Yeah I like that idea. We can just use the merkle trie key order to sort things, no need for a priority heap. :)

Perhaps we could even round priorities to `u16` if we want 4 instead of 16 trie nodes per receipts. (~16k levels of priority seem enough to me, but I could be wrong)
That would also make attacks more expensive, as the priority increasing steps would need to be higher.

But one question. I'm reading your proposal as the key being `TrieKey::PriorityReceipts{priority: u64}`. But what is the value? I guess `{ receipts: Vec<CryptoHash> }` would make sense, storing receipt ids to lookup the value in the delayed queue. Is that what you have in mind?

Reply:
Value is a `64` (index of delayed receipt)